### PR TITLE
Fix bug in eval

### LINKED ISSUE #288: Fix eval bug
RV crashes when evaluating on scenes where not all classes are present in the ground truth.

### PR DESCRIPTION
**Pull request by lewfish**
This PR fixes a bug in the eval command and adds some extra refactoring and documentation. Running eval on the problematic xView scene should now work.

Closes #288

### CODE REVIEW

**Comment by simonkassel** [comment]
I'm still getting an error when I try to run this on xView

```Traceback (most recent call last):
  File "/usr/lib/python3.5/runpy.py", line 184, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/opt/src/rastervision/utils/chain_workflow.py", line 450, in <module>
    main()
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.5/dist-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/opt/src/rastervision/utils/chain_workflow.py", line 446, in main
    workflow.local_run(tasks)
  File "/opt/src/rastervision/utils/chain_workflow.py", line 421, in local_run
    run._eval(self.path_generator.eval_config_uri)
  File "/opt/src/rastervision/run.py", line 54, in _eval
    command.run()
  File "/opt/src/rastervision/commands/eval.py", line 12, in run
    self.ml_task.eval(self.scenes, self.options)
  File "/opt/src/rastervision/core/ml_task.py", line 195, in eval
    self.class_map, ground_truth, predictions)
  File "/opt/src/rastervision/evaluations/object_detection_evaluation.py", line 89, in compute
    self.class_to_eval_item = parse_od_eval(od_eval, class_map)
  File "/opt/src/rastervision/evaluations/object_detection_evaluation.py", line 59, in parse_od_eval
    precision = float(precisions[-1])
IndexError: index -1 is out of bounds for axis 0 with size 0
```

`od_eval.precisions_per_class` seems to be full of empty arrays. When I print it out:

```[array([], dtype=float64), array([], dtype=float64), array([], dtype=float64), array([], dtype=float64), array([], dtype=float64), array([], dtype=float64), array([], dtype=float64), array([], dtype=float64), array([], dtype=float64)]```

this may be specific to my case?

**Comment by simonkassel** [comment]
I got this to run on the xview dataset. It looks good. I think I was having a distinct issue. I fixed it by only evaluating for a class if it has a non-empty precisions array. 

@lewfish do you want me to submit a PR merging into this branch or should we handle it separately?

File: src/rastervision/evaluations/object_detection_evaluation.py [hunk 1]
```diff
@@ -29,15 +27,31 @@ def compute_od_eval(ground_truth_labels, prediction_labels):
...
+            # precision_per_class can be shorter than the total number of
+            # classes in the class_map. Therefore, we use score_ind to index
+            # into precisions_per_class instead of simply using class_id - 1.
+            score_ind += 1
+
+            # Precisions and recalls across a range of detection thresholds.
+            precisions = od_eval.precisions_per_class[score_ind]
+            recalls = od_eval.recalls_per_class[score_ind]
+
+            # If we use the lowest detection threshold (ie. use all detected
```
Comment (simonkassel) [inline src/rastervision/evaluations/object_detection_evaluation.py]: to get this to run with xview i had to change this only run the code below if precisions and recalls were not empty arrays

Comment (lewfish) [reply src/rastervision/evaluations/object_detection_evaluation.py]: You can add a commit to this one, and I'll take a look at it.

Comment (simonkassel) [reply src/rastervision/evaluations/object_detection_evaluation.py]: ok, just did

**Comment by lewfish** [comment]
The commit you added is to guard against the case where there are no predictions for the scene. It's not quite correct though since we should still add an eval item (with a recall of zero) when this occurs. I'm going to fix this as part of the eval unit tests I'm adding, so I'm going to remove the commit you added and then merge.

**Pull request merged**